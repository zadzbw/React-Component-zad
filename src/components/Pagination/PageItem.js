/**
 * Created by zad on 17/2/3.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import {paginationPrefix} from './Pagination';

export default class PageItem extends Component {
  static propTypes = {
    page: PropTypes.number,
    active: PropTypes.bool,
    itemClick: PropTypes.func,
  };

  render() {
    const {page, active, itemClick} = this.props;
    const itemClass = classNames(`${paginationPrefix}-item`, {
      [`${paginationPrefix}-item-active`]: active,
    });

    return (
      <li title={page} className={itemClass} onClick={itemClick}>
        <a>{page}</a>
      </li>
    );
  }
}
