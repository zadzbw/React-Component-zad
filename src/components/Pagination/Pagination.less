@import "../../style/variables";
@import "../../style/mixins/index";

.zad-pagination {
  font-size: @font-size-base;
  .clearfix();

  &-item {
    float: left;
    cursor: pointer;
    border-radius: @border-radius-base;
    user-select: none;
    min-width: 28px;
    height: 28px;
    line-height: 26px;
    text-align: center;
    list-style: none;
    border: 1px solid @border-color-base;
    background-color: @body-background;
    margin-right: 8px;

    a {
      text-decoration: none;
      color: @text-color;
      margin: 0 6px;
      transition: none; // 在切换页数的时候，去除动画
    }

    &:hover {
      transition: all @animation-duration-slow ease;
      border-color: @primary-color;
      a {
        color: @primary-color;
      }
    }

    &-active {
      border-color: @primary-color;
      background-color: @primary-color;
      &:hover a, a {
        color: @body-background;
      }
    }
  }

  &-prev, &-next, &-jump-prev, &-jump-next {
    float: left;
    position: relative;
    cursor: pointer;
    border-radius: @border-radius-base;
    color: @text-color;
    user-select: none;
    min-width: 28px;
    height: 28px;
    line-height: 26px;
    text-align: center;
    list-style: none;
    margin-right: 8px;
    transition: all @animation-duration-slow ease;
    display: inline-block;
  }

  &-jump-prev, &-jump-next {
    a {
      &:after {
        content: "•••";
        display: block;
        color: @disabled-color;
        text-align: center;
      }
    }

    &:hover {
      a {
        &:after {
          font: normal normal normal 14px/1 FontAwesome;
          color: @primary-color;
          position: absolute;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
          display: block;
          letter-spacing: 1px;
        }
      }
    }
  }

  &-jump-prev {
    &:hover {
      a {
        &:after {
          content: "\f104\f104";
        }
      }
    }
  }

  &-jump-next {
    &:hover {
      a {
        &:after {
          content: "\f105\f105";
        }
      }
    }
  }

  &-prev, &-next {
    border: 1px solid @border-color-base;
    background-color: @body-background;

    a {
      color: @text-color;
      &:after {
        font: normal normal normal 14px/1 FontAwesome;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        display: block;
      }
    }

    &:hover {
      border-color: @primary-color;
      a {
        color: @primary-color;
      }
    }
  }

  &-prev {
    a {
      &:after {
        content: "\f104";
      }
    }
  }

  &-next {
    a {
      &:after {
        content: "\f105";
      }
    }
  }

  &-disabled {
    &:hover {
      border-color: @border-color-base;
      a {
        color: @disabled-color;
        cursor: not-allowed;
      }
    }
    cursor: not-allowed;
    a {
      color: @disabled-color;
    }
  }

  &-quick-jumper {
    float: left;
    height: 28px;
    line-height: 26px;
    user-select: none;

    .zad-input {
      width: 50px;
      margin: 0 8px;
    }
  }
}
