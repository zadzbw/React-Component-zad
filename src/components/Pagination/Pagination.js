/**
 * Created by zad on 17/1/20.
 */
import React, {PropTypes, Component} from 'react';
import classNames from 'classnames';
import PageItem from './PageItem';

export const paginationPrefix = 'zad-pagination';

export default class Pagination extends Component {
  constructor(props) {
    super(props);
    this.state = {
      _current: 'current' in this.props ? this.props.current : this.props.defaultCurrent,
      _pageSize: 'selectedKeys' in this.props ? this.props.pageSize : this.props.defaultPageSize,
    };
    this._getPage = this::this._getPage;
    this._isValid = this::this._isValid;
    this._onPageChange = this::this._onPageChange;
    this._hasPrev = this::this._hasPrev;
    this._hasNext = this::this._hasNext;
    this._clickPrev = this::this._clickPrev;
    this._clickNext = this::this._clickNext;
    this._jumpPrev = this::this._jumpPrev;
    this._jumpNext = this::this._jumpNext;
    this._quickJump = this::this._quickJump;
  }

  // todo 等把Select组件完成后，加入pageSizeChange功能

  static propTypes = {
    current: PropTypes.number,
    defaultCurrent: PropTypes.number,
    total: PropTypes.number,
    pageSize: PropTypes.number,
    defaultPageSize: PropTypes.number,
    onChange: PropTypes.func,
    showQuickJumper: PropTypes.bool,
    onPageSizeChange: PropTypes.func,
    showSizeChanger: PropTypes.bool,
    pageSizeOption: PropTypes.arrayOf(PropTypes.number),
  };

  static defaultProps = {
    defaultCurrent: 1,
    total: 0,
    defaultPageSize: 10,
    onChange: (page) => undefined,
    showQuickJumper: false,
    onPageSizeChange: (size) => undefined,
    showSizeChanger: false,
    pageSizeOption: [10, 20, 30, 40],
  };

  componentWillReceiveProps(nextProps) {
    if ('current' in nextProps) {
      this.setState({
        _current: nextProps.current,
      });
    }
  }

  _getPage(p) {
    let _pageSize = p;
    if (typeof _pageSize === 'undefined') {
      _pageSize = this.state._pageSize;
    }
    return Math.ceil(this.props.total / _pageSize);
  }

  _isValid(page) {
    return typeof page === 'number' && page >= 1 && page !== this.state._current;
  }

  _onPageChange(page) {
    if (this._isValid(page)) {
      page = page > this._getPage() ? this._getPage() : page;
      if (!('current' in this.props)) {
        this.setState({
          _current: page,
        });
      }
      this.props.onChange(page);
    }
  }

  _hasPrev() {
    return this.state._current > 1;
  }

  _hasNext() {
    return this.state._current < this._getPage();
  }

  _clickPrev() {
    this._hasPrev() && this._onPageChange(this.state._current - 1);
  }

  _clickNext() {
    this._hasNext() && this._onPageChange(this.state._current + 1);
  }

  _jumpPrev() {
    this._onPageChange(Math.max(1, this.state._current - 5));
  }

  _jumpNext() {
    this._onPageChange(Math.min(this._getPage(), this.state._current + 5));
  }

  _quickJump(e) {
    if (e.keyCode === 13) {
      const {total} = this.props;
      const {_current} = this.state;
      const inputValue = e.target.value;
      const maxPage = this._getPage();

      if (inputValue.length && total !== 0) {
        if (isNaN(inputValue)) {
          e.target.value = _current;
          console.warn('Please input a Number');
        } else {
          if (1 <= inputValue && inputValue <= maxPage) {
            if (_current !== Math.round(inputValue)) {
              this._onPageChange(Math.round(inputValue));
              e.target.value = Math.round(inputValue);
            }
          } else if (inputValue < 1) {
            this._onPageChange(1);
            e.target.value = 1;
          } else {
            this._onPageChange(maxPage);
            e.target.value = maxPage;
          }
        }
      } else {
        e.target.value = '';
      }
    }
  }

  render() {
    const {showQuickJumper}= this.props;
    const {_current} = this.state;
    const pages = this._getPage();
    const itemList = [];

    const prevClass = classNames(`${paginationPrefix}-prev`,
      {[`${paginationPrefix}-disabled`]: !this._hasPrev()}
    );
    const nextClass = classNames(`${paginationPrefix}-next`,
      {[`${paginationPrefix}-disabled`]: !this._hasNext()}
    );

    if (pages <= 9) {
      for (let i = 1; i <= pages; i++) {
        itemList.push(
          <PageItem
            key={i}
            page={i}
            active={i === _current}
            itemClick={this._onPageChange.bind(this, i)}
          />
        );
      }
    } else {
      let left = Math.max(1, _current - 2);
      let right = Math.min(_current + 2, pages);
      if (_current <= 3) {
        right = 5; // 1 + 4
      }
      if (pages - _current <= 2) {
        left = pages - 4;
      }
      for (let i = left; i <= right; i++) {
        itemList.push(
          <PageItem
            key={i}
            page={i}
            active={i === _current}
            itemClick={this._onPageChange.bind(this, i)}
          />
        );
      }

      // 向前或向后跳转
      if (_current >= 4) {
        itemList.unshift(
          <li
            className={`${paginationPrefix}-jump-prev`}
            key="to-prev"
            title="向前5页"
            onClick={this._jumpPrev}
          >
            <a/>
          </li>
        );
      }
      if (_current <= pages - 3) {
        itemList.push(
          <li
            className={`${paginationPrefix}-jump-next`}
            key="to-next"
            title="向后5页"
            onClick={this._jumpNext}
          >
            <a/>
          </li>
        );
      }

      // 跳转至首页与尾页
      if (left > 1) {
        itemList.unshift(
          <PageItem
            key={1}
            page={1}
            active={false}
            itemClick={this._onPageChange.bind(this, 1)}
          />
        );
      }
      if (right < pages) {
        itemList.push(
          <PageItem
            key={pages}
            page={pages}
            active={false}
            itemClick={this._onPageChange.bind(this, pages)}
          />
        );
      }
    }

    const quickJumper = !showQuickJumper ? null : (
      <li className={`${paginationPrefix}-quick-jumper`}>
        to
        <input type="text" onKeyUp={this._quickJump} defaultValue={_current}/>
      </li>
    );

    return (
      <ul className={paginationPrefix}>
        <li
          title={'上一页'}
          className={prevClass}
          onClick={this._clickPrev}
        >
          <a/>
        </li>
        {itemList}
        <li
          title={'下一页'}
          className={nextClass}
          onClick={this._clickNext}
        >
          <a/>
        </li>
        {quickJumper}
      </ul>
    );
  }
}
