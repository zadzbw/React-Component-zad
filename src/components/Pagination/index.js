/**
 * Created by zad on 17/1/20.
 */
import Pagination from './Pagination';
import './Pagination.less';

export default Pagination;
