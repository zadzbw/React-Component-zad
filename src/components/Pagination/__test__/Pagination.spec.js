/**
 * Created by zad on 17/2/6.
 */
import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import Pagination from '../../Pagination';
import PageItem from '../../Pagination/PageItem';

describe('Pagination Test', () => {
  it('Normal', () => {
    const wrapper = shallow(
      <Pagination defaultCurrent={3} total={155}/>
    );

    // page 3
    expect(wrapper.find('.zad-pagination-jump-prev').isEmpty()).toBeTruthy();
    expect(wrapper.find('.zad-pagination-jump-next').isEmpty()).toBeFalsy();
    expect(wrapper.state()._current).toEqual(3);
    expect(wrapper.find(PageItem)).toHaveLength(6);
    expect(wrapper.find('.zad-pagination-prev').hasClass('zad-pagination-disabled')).toBeFalsy();
    expect(wrapper.find('.zad-pagination-next').hasClass('zad-pagination-disabled')).toBeFalsy();

    // page 8
    wrapper.shallow().find('.zad-pagination-jump-next').at(0).shallow().simulate('click');
    expect(wrapper.find('.zad-pagination-jump-prev').isEmpty()).toBeFalsy();
    expect(wrapper.find('.zad-pagination-jump-next').isEmpty()).toBeFalsy();
    expect(wrapper.state()._current).toEqual(8);
    expect(wrapper.find(PageItem)).toHaveLength(7);
    expect(wrapper.find('.zad-pagination-prev').hasClass('zad-pagination-disabled')).toBeFalsy();
    expect(wrapper.find('.zad-pagination-next').hasClass('zad-pagination-disabled')).toBeFalsy();

    // page last
    wrapper.shallow().find(PageItem).at(6).shallow().simulate('click');
    expect(wrapper.find('.zad-pagination-jump-prev').isEmpty()).toBeFalsy();
    expect(wrapper.find('.zad-pagination-jump-next').isEmpty()).toBeTruthy();
    expect(wrapper.state()._current).toEqual(16);
    expect(wrapper.find(PageItem)).toHaveLength(6);
    expect(wrapper.find('.zad-pagination-prev').hasClass('zad-pagination-disabled')).toBeFalsy();
    expect(wrapper.find('.zad-pagination-next').hasClass('zad-pagination-disabled')).toBeTruthy();

    // page first
    wrapper.shallow().find(PageItem).at(0).shallow().simulate('click');
    expect(wrapper.find('.zad-pagination-jump-prev').isEmpty()).toBeTruthy();
    expect(wrapper.find('.zad-pagination-jump-next').isEmpty()).toBeFalsy();
    expect(wrapper.state()._current).toEqual(1);
    expect(wrapper.find(PageItem)).toHaveLength(6);
    expect(wrapper.find('.zad-pagination-prev').hasClass('zad-pagination-disabled')).toBeTruthy();
    expect(wrapper.find('.zad-pagination-next').hasClass('zad-pagination-disabled')).toBeFalsy();

    expect(wrapper.find('.zad-pagination-quick-jumper').isEmpty()).toBeTruthy();
    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });

  it('QuickJump', () => {
    const wrapper = shallow(
      <Pagination defaultCurrent={3} total={155} showQuickJumper/>
    );
    expect(wrapper.find('.zad-pagination-quick-jumper').isEmpty()).toBeFalsy();

    // to page 9
    wrapper.find('.zad-pagination-quick-jumper').find('input').simulate('keyUp', {keyCode: 13, target: {value: '9'}});
    expect(wrapper.find('.zad-pagination-jump-prev').isEmpty()).toBeFalsy();
    expect(wrapper.find('.zad-pagination-jump-next').isEmpty()).toBeFalsy();
    expect(wrapper.state()._current).toEqual(9);
    expect(wrapper.find(PageItem)).toHaveLength(7);
    expect(wrapper.find('.zad-pagination-prev').hasClass('zad-pagination-disabled')).toBeFalsy();
    expect(wrapper.find('.zad-pagination-next').hasClass('zad-pagination-disabled')).toBeFalsy();

    expect(shallowToJson(wrapper.shallow())).toMatchSnapshot();
  });
});
