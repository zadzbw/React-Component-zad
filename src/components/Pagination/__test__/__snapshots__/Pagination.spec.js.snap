exports[`Pagination Test Normal 1`] = `
<ul
  className="zad-pagination">
  <li
    className="zad-pagination-prev zad-pagination-disabled"
    onClick={[Function]}
    title="上一页">
    <a />
  </li>
  <PageItem
    active={true}
    itemClick={[Function]}
    page={1} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={2} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={3} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={4} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={5} />
  <li
    className="zad-pagination-jump-next"
    onClick={[Function]}
    title="向后5页">
    <a />
  </li>
  <PageItem
    active={false}
    itemClick={[Function]}
    page={16} />
  <li
    className="zad-pagination-next"
    onClick={[Function]}
    title="下一页">
    <a />
  </li>
</ul>
`;

exports[`Pagination Test QuickJump 1`] = `
<ul
  className="zad-pagination">
  <li
    className="zad-pagination-prev"
    onClick={[Function]}
    title="上一页">
    <a />
  </li>
  <PageItem
    active={false}
    itemClick={[Function]}
    page={1} />
  <li
    className="zad-pagination-jump-prev"
    onClick={[Function]}
    title="向前5页">
    <a />
  </li>
  <PageItem
    active={false}
    itemClick={[Function]}
    page={7} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={8} />
  <PageItem
    active={true}
    itemClick={[Function]}
    page={9} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={10} />
  <PageItem
    active={false}
    itemClick={[Function]}
    page={11} />
  <li
    className="zad-pagination-jump-next"
    onClick={[Function]}
    title="向后5页">
    <a />
  </li>
  <PageItem
    active={false}
    itemClick={[Function]}
    page={16} />
  <li
    className="zad-pagination-next"
    onClick={[Function]}
    title="下一页">
    <a />
  </li>
  <li
    className="zad-pagination-quick-jumper">
    to
    <input
      defaultValue={9}
      onKeyUp={[Function]}
      type="text" />
  </li>
</ul>
`;
